### DISCUSSION TITLE
paketo buildpacks maven: download dependencies behind corporate proxy

### DISCUSSION THREAD

**Discussion by NValanos** [paketo-buildpacks/feedback#318] [Dependencies Team]
Hi All,

I have a simple spring boot project which I try to build through gitlab, using .gitlab-ci.yml . I have the limitation that I need to run behind a docker container, so decided to use the `paketobuildpacks/builder-jammy-full` image and run the building process using the command: `/cnb/lifecycle/creator`. The process runs smoothly (I have set the proxy in the .docker/config.json) and it downloads any required image, but when the paketobuildpacks/maven pack runs, after it successfully downloads maven from repo, when running
```mvn --settings=settings.xml --batch-mode -Dmaven.test.skip=true --no-transfer-progress package```
it is unable to download the spring boot parent image from the repository. The error is the following:

```
Could not transfer artifact org.springframework.boot:spring-boot-starter-parent:pom:3.3.4 from/to central (https://repo.maven.apache.org/maven2): Network is unreachable and 'parent.relativePath' points at no local POM @ line 5, column 10 -> [Help 2]
```

The reason is probably due to the fact that maven does not use the http_proxy. The possible solution would be to set the proxy inside the  settings.xml file. But even though I set the location of the settings.xml file through the environmental variable `BP_MAVEN_SETTINGS_PATH=setting.xml` (It is the same path as the one set by default for the pom.xml. It is basically the root of the project.) I get an error: 

```
Executing mvn --settings=settings.xml
     --batch-mode -Dmaven.test.skip=true --no-transfer-progress package
      [ERROR] Error executing Maven.
      [ERROR] The specified user settings file does not exist: /<path-to-project>/settings.xml
```

I have already checked and the settings.xml exists in the specified folder. Also, I have tried changing the permissions of the file just in case, but with no success.

Could there be any specific folder that I have to copy the settings.xml in order to readable by maven? If you need more information about the settings of the environment, I would be glad to provide. Thank you in advance.

P.S. Some additional information: the spring boot project does not contain any .mvn folder or mvnw files. Also no bindings have been currently set in the pom.xml (might be needed in the future)

**Comment by dmikusa** [comment]
There shouldn't be any specific place you need to put this. When no bindings are set, the path that you set in `BP_MAVEN_SETTINGS_PATH` is set as the argument `--settings=<path>`. No other adjustments or changes are made to the value you specify.

Only things off the top of my head:

1. Make sure the path you specify exists in the container where Maven runs. It needs to be the path as Maven will see it. Your case is slightly different because the container where your build runs is the CI build environment itself and then you invoke `creator`. Most users will use Pack or Spring Boot build tools which will set up a container for them.
2. Maybe try a full path to the file in `BP_MAVEN_SETTINGS_PATH`. I don't think it should matter so long as the relative path is correct, but it's worth a shot.

If the file exists, has the permissions for the user running Maven to read the file, and you give Maven the correct path to the file, I don't see any reason why Maven would be unable to read it and use the values there.

**Reply by NValanos** [reply]
Hello again,

Thanks for the response! 

The use of the `creator` is required because we need to run in a dockerless container environment for the sake of security. That's why we decided to use `paketobuildpacks/builder-jammy-full` as a base.

I run some commands inside the container through gitlab-ci in order to determine the location of the file. From the output of the commands, it seems that settings.xml exists in the location specified. For the sake of testing, I also changed the path from relative to absolute, with no success. I provide some logs in case they may help:

```
$ whoami
root
$ export CNB_PLATFORM_API=0.14
$ pwd
/builds/cicd-tests/tests/ci-test
$ ls -la
total 28
drwxrwxrwx 6 root root  164 Oct 29 09:59 .
drwxrwxrwx 4 root root   40 Oct 29 08:39 ..
drwxrwxrwx 6 root root  128 Oct 29 09:59 .git
-rw-rw-rw- 1 root root  280 Oct 29 08:39 .gitignore
-rw-rw-rw- 1 root root 1722 Oct 29 09:59 .gitlab-ci.yml
drwxrwxrwx 3 root root   21 Oct 29 08:39 .mvn
-rw-rw-rw- 1 root root 1031 Oct 29 08:39 HELP.md
-rw-rw-rw- 1 root root 6191 Oct 29 08:39 README.md
drwxrwxrwx 3 root root   29 Oct 29 08:42 bindings
-rw-rw-rw- 1 root root 3037 Oct 29 08:47 pom.xml
-rw-rw-rw- 1 root root 2215 Oct 29 09:38 settings.xml
drwxrwxrwx 4 root root   30 Oct 29 08:39 src
$ echo "/builds/cicd-tests/tests/ci-test/settings.xml" > /platform/env/BP_MAVEN_SETTINGS_PATH
$ /cnb/lifecycle/creator -app=. $CI_REGISTRY_IMAGE:latest

... (the proccess runs smoothly, until)

  Apache Maven 3.9.9: Contributing to layer
    Downloading from https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.9.9/apache-maven-3.9.9-bin.tar.gz
    Verifying checksum
    Expanding to /layers/paketo-buildpacks_maven/maven
    Creating cache directory /home/cnb/.m2
  Compiled Application: Contributing to layer
    Executing mvn --settings=/builds/cicd-tests/tests/ci-test/settings.xml
     --batch-mode -Dmaven.test.skip=true --no-transfer-progress package
      [ERROR] Error executing Maven.
      [ERROR] The specified user settings file does not exist: /builds/cicd-tests/tests/ci-test/settings.xml
      
      unable to invoke layer creator
unable to contribute application layer
error running build
exit status 1
ERROR: failed to build: exit status 1

```

It is also worth mentioning that I run the container as *root* user. Don't know if this has anything to do with it though.
The `ls -la` command shows the spring boot project folder. The `bindings` folder contains some required configuration for certificates of the project. Nothing relevant to maven.

We appreciate any advice. Thanks again for your time.

**Reply by dmikusa** [reply]
> Executing mvn --settings=/builds/cicd-tests/tests/ci-test/settings.xml
>     --batch-mode -Dmaven.test.skip=true --no-transfer-progress package

This output is coming from Maven itself. I would suggest that you try and run Maven manually and see if you can get it to find the file. That would help further break down the issue. After you run the buildpacks once, you should see it under `/layers/paketo-buildpacks_maven/maven`. 

> It is also worth mentioning that I run the container as root user.

That's not something you'd typically do with buildpacks, so I suppose it's possible that it could be causing issues but I don't know of anything specific off the top of my head. I see that your settings file is rw for world so it should be visible regardless of the permissions. 

You might also try stepping through: https://medium.com/buildpacks/unpacking-cloud-native-buildpacks-ff51b5a767bf to see if there are any steps that you may have missed in the set up of the build environment.

**Reply by NValanos** [reply]
I finally found the answer. Due to the fact that I was using the image as root (I needed permission in order to include some certificates), the 'settings.xml' was in a different user/group from the one that the builder had access to.

By creating the folder `/home/cnb/.m2` (which is a folder that is automatically created by the builder during the maven package run), copying the `settings.xml` inside the folder and then changing the owner and group to `cnb` (the user and group used by the build), maven was able to read the configuration, even without explicitly setting the environmental variable `BP_MAVEN_SETTINGS_PATH`).

For any future reference, I include the steps as I used them inside my gitlab-ci.yml file:
```
    - mkdir /home/cnb/.m2
    - cp ./settings.xml /home/cnb/.m2/ # settings.xml should already exist in the application folder, thus why I include it using the relative default path that the container starts into
    - chown -R cnb:cnb /home/cnb/.m2/
```

Thank you again for your support. I'm closing the discussion as resolved
